Lib/asyncio/base_events.py

```diff
@@ -844,12 +879,49 @@ def _check_sendfile_params(self, sock, file, offset, count):
                 "offset must be a non-negative integer (got {!r})".format(
                     offset))
 
+    async def _connect_sock(self, exceptions, addr_info, local_addr_infos=None):
+        """Create, bind and connect one socket."""
+        my_exceptions = []
+        exceptions.append(my_exceptions)
+        family, type_, proto, _, address = addr_info
+        sock = None
+        try:
+            sock = socket.socket(family=family, type=type_, proto=proto)
+            sock.setblocking(False)
+            if local_addr_infos is not None:
+                for _, _, _, _, laddr in local_addr_infos:
+                    try:
+                        sock.bind(laddr)
+                        break
+                    except OSError as exc:
+                        msg = (
+                            f'error while attempting to bind on '
+                            f'address {laddr!r}: '
+                            f'{exc.strerror.lower()}'
+                        )
+                        exc = OSError(exc.errno, msg)
+                        my_exceptions.append(exc)
+                else:  # all bind attempts failed
+                    raise my_exceptions.pop()
+            await self.sock_connect(sock, address)
+            return sock
+        except OSError as exc:
+            my_exceptions.append(exc)
+            if sock is not None:
+                sock.close()
+            raise
+        except:
+            if sock is not None:
+                sock.close()
+            raise
+
     async def create_connection(
             self, protocol_factory, host=None, port=None,
             *, ssl=None, family=0,
             proto=0, flags=0, sock=None,
             local_addr=None, server_hostname=None,
-            ssl_handshake_timeout=None):
+            ssl_handshake_timeout=None,
+            delay=None, interleave=None):
```

Comment:
I think that `delay` parameter doesn't clearly convey that it switches on a completely different connect behaviour. Maybe call it "happy_eyeballs_delay"?  

@asvetlov what do you think?  BTW, do you guys have happy eyeballs implementation in aiohttp?

Reply:
No, we have not (yet). We iterate over available hosts one by one.
BTW there is potentially related thing: SRV DNS records.
It can be useful for organizing load balancing and high available systems.
Rust supports it, would be nice to have it in asyncio too.

Reply:
I prefer `happy_eyeballs: HappyEyeBallsConfig=None` where `HappyEyeBallsConfig` is a dataclass with all required configuration parameters.

Reply:
> I prefer happy_eyeballs: HappyEyeBallsConfig=None where HappyEyeBallsConfig is a dataclass with all required configuration parameters.

Oh, that's a bit too heavy IMO. We don't have a precedent for such API design neither in asyncio nor in any other popular APIs. And maybe `interleave` isn't a happy eyeballs specific thing?

I think that `interleave=True` is meaningful even when happy eyeballs isn't enabled:

```python
await loop.create_connection(proto_factory, "google.com", 80, interleave=True)
# connect to google, but shuffle the addresses that getaddrinfo resolves to
```

If that's the case, then we want just `happy_eyeballs_delay` parameter to enable happy eyeballs algorithm.

Reply:
> BTW there is potentially related thing: SRV DNS records.
> It can be useful for organizing load balancing and high available systems.
> Rust supports it, would be nice to have it in asyncio too.

I don't know how prevalent SRV is used, but since the operating system's `getaddrinfo` doesn't support anything but A and AAAA queries, some other DNS library will be required to do SRV queries. (Linux has `libresolv`, not sure whether it's cross platform.) At that point it's probably too big to fit in `asyncio`.